### PR TITLE
Improve PF UI

### LINKED ISSUE #2463: Improve payload formatters UI in the Console
#### Summary

I think it could be beneficial to change UI of payload formatters in the Console.

References https://github.com/TheThingsNetwork/lorawan-stack/issues/2232

#### Why do we need this?

Better user experience. Currently, it is quite difficult to determine which formatter is used for a particular end device when both app/device formatters are set. Also, payload formatters options are not documented anywhere which is confusing (`None` vs `Repository`?). Additionally, refactoring the code for payload formatters (in the Console) would be nice as well.

#### What is already there? What do you see now?

Screenshot 2020-04-30 at 16 49 46

#### What is missing? What do you want to see?

1. Inform the user when on the end device payload formatters page if there is app formatter set and which
2. Inform the user which payload formatter is used for a particular device
3. Overall improved UI
4. Separate UI's for app/device payload formatters pages

#### Environment

v3.7.2

#### Can you do this yourself and submit a Pull Request?

yes

@laurensslats  @benolayinka  - documentation, suggestions, feedback
@kschiffer  @pierrephz - design, suggestions

cc @johanstokking

### PR DESCRIPTION
**Pull request by ryaplots**
#### Summary

- Closes [#691](https://github.com/TheThingsIndustries/lorawan-stack-support/issues/691)
- Closes #2463 

#### Changes

- Add paste buttons
Screenshot 2022-01-24 at 12 40 33
Screenshot 2022-01-24 at 12 40 46

- Change notification position and add information about the app PF
Screenshot 2022-01-24 at 12 40 59
Screenshot 2022-01-24 at 12 41 32

#### Testing

e2e tests and manually

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/webui/components/notification/index.js [hunk 1]
```diff
@@ -66,7 +67,7 @@ const Notification = ({
               content={content}
               values={messageValues}
               firstToUpper
-              convertBackticks={Boolean(error)}
+              convertBackticks={Boolean(error) || convertBackticks}
```
Comment (kschiffer) [inline pkg/webui/components/notification/index.js]: Let's actually just enable this by default in the `<Message />` component. I think that will save us a lot of hassle of propagating the prop through all the time.

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 2]
```diff
@@ -43,15 +54,20 @@ const m = defineMessages({
   formatterType: 'Formatter type',
   formatterParameter: 'Formatter parameter',
   grpcFieldDescription: 'The address of the service to connect to',
-  appFormatter: 'Use application payload formatter',
+  appFormatter: 'Use application payload formatter  ',
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
  appFormatter: 'Use application payload formatter',
```

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 3]
```diff
@@ -86,10 +102,29 @@ const validationSchema = Yup.object().shape({
...
+    }
+  },
+  dispatch => ({
+    getRepositoryPayloadFormatters: (appId, versionIds) =>
+      dispatch(getRepositoryPayloadFormatters(appId, versionIds)),
+  }),
+)
+@withRequest(({ appId, versionIds, getRepositoryPayloadFormatters }) =>
+  getRepositoryPayloadFormatters(appId, versionIds),
+)
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]: If you connect the component to the state, it will [no longer be a regular component but a container component](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.18/DEVELOPMENT.md#react-component-types). It should the be moved to the `container/` folder. I think it's better though to create a separate container that takes care of fetching and then passes down the data as props to the form component.

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 4]
```diff
@@ -212,23 +279,58 @@ class PayloadFormattersForm extends React.Component {
...
-        return null
+          {type === TYPES.JAVASCRIPT && (
+            <>
+              <Button
+                type="button"
+                message={m.pasteApplicationFormatter}
+                secondary
+                icon={'payload_formats'}
+                onClick={pasteApplicationFormatter}
+              />
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]: Keep in mind that the application PF can be set to `None`. In that case, the button should be hidden.

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 5]
```diff
@@ -212,23 +279,58 @@ class PayloadFormattersForm extends React.Component {
...
-        )
-      default:
-        return null
+          {type === TYPES.JAVASCRIPT && (
+            <>
+              <Button
+                type="button"
+                message={m.pasteApplicationFormatter}
+                secondary
+                icon={'payload_formats'}
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
```
Better without icon.

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 6]
```diff
@@ -212,23 +279,58 @@ class PayloadFormattersForm extends React.Component {
...
+                onClick={pasteApplicationFormatter}
+              />
+              {Boolean(repositoryPayloadFormatters) && (
+                <Button
+                  type="button"
+                  message={m.pasteRepositoryFormatter}
+                  secondary
+                  icon={'payload_formats'}
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
```

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 7]
```diff
@@ -212,23 +279,58 @@ class PayloadFormattersForm extends React.Component {
...
+          type="text"
+          placeholder={sharedMessages.addressPlaceholder}
+          description={m.grpcFieldDescription}
+          autoComplete="on"
+        />
+      )
+    } else if (showRepositoryParameter) {
+      return (
+        <Form.Field
+          readOnly={isReadOnly}
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
          readOnly
```

?

File: pkg/webui/console/containers/device-payload-formatters/downlink.js [hunk 8]
```diff
@@ -88,10 +86,12 @@ class DevicePayloadFormatters extends React.PureComponent {
       }),
     }).isRequired,
     updateDevice: PropTypes.func.isRequired,
+    version_ids: PropTypes.shape({}),
```
Comment (kschiffer) [inline pkg/webui/console/containers/device-payload-formatters/downlink.js]: I think this can be defined with at least the props that the console expects.

File: pkg/webui/console/containers/device-payload-formatters/downlink.js [hunk 9]
```diff
@@ -60,6 +57,7 @@ import messages from './messages'
       link: selectApplicationLink(state),
       formatters,
       encodeDownlink: tts.As.encodeDownlink,
+      version_ids: selectVersionIds(state),
```
Comment (kschiffer) [inline pkg/webui/console/containers/device-payload-formatters/downlink.js]: The new state props can be retrieved here and then passed on. See my comment above on moving state logic to the container prop.

File: pkg/webui/console/store/middleware/logics/device-repository.js [hunk 10]
```diff
@@ -84,10 +84,44 @@ const getTemplateLogic = createRequestLogic({
...
+        version,
+      )
+      const repositoryPayloadFormatters = {
+        ...uplinkDecoder,
+        ...downlinkDecoder,
+        ...downlinkEncoder,
+      }
+      return repositoryPayloadFormatters
+    } catch (error) {
+      // There is no codec for this end device.
```
Comment (kschiffer) [inline pkg/webui/console/store/middleware/logics/device-repository.js]: I believe it's also possible to have only one the codecs. Like this, no PF is recognized if only one is not existing.

Also, the catch is too general, there could be network or server errors that do not mean that the codecs do not exist. You'll probably want to check for `isNotFoundError(error)` only and rethrow any other error.

File: pkg/webui/console/store/middleware/logics/devices.js [hunk 11]
```diff
@@ -34,7 +34,7 @@ const getDeviceLogic = createRequestLogic({
     } = action
     const dev = await tts.Applications.Devices.getById(appId, deviceId, selector)
     dispatch(devices.startDeviceEventsStream(dev.ids))
-    return dev
+    return { ...dev, version_ids: dev.version_ids }
```
Comment (kschiffer) [inline pkg/webui/console/store/middleware/logics/devices.js]: Remove this change.

File: pkg/webui/console/store/reducers/devices.js [hunk 12]
```diff
@@ -66,10 +67,11 @@ const devices = (state = defaultState, { type, payload, event }) => {
       return {
         ...state,
         selectedDevice: combineDeviceIds(payload.appId, payload.deviceId),
+        version_ids: payload.version_ids,
       }
     case UPDATE_DEV_SUCCESS:
     case GET_DEV_SUCCESS:
-      const updatedState = { ...state }
+      const updatedState = { ...state, version_ids: payload.version_ids }
```
Comment (kschiffer) [inline pkg/webui/console/store/reducers/devices.js]: I don't get these changes. The `version_ids` should be part of the device entity stored in the store.

**Review by kschiffer — CHANGES_REQUESTED** [review]
Review is WIP but here are some change requests already.

Functionality-wise it's working well.

File: pkg/webui/console/store/middleware/logics/device-repository.js [hunk 13]
```diff
@@ -104,15 +106,18 @@ const getRepositoryPayloadFormattersLogic = createRequestLogic({
...
       }
       return repositoryPayloadFormatters
     } catch (error) {
-      // There is no codec for this end device.
-      return null
+      if (isNotFoundError(error) && typeof repositoryPayloadFormatters !== 'undefined') {
+        return repositoryPayloadFormatters
+      }
+
+      throw error
```
Comment (kschiffer) [inline pkg/webui/console/store/middleware/logics/device-repository.js]: Still not happy with this logic:

1. It runs three requests successively that could be run in parallel
2. If the first request errors out with a 404, the other requests will also not run (a thrown exception will stop further execution of the current `try`-block)

The if condition here can never be true because the error would occur before the assignment of `repositoryPayloadFormatters`

You will likely want something like this:
```js
  const ignoreNotFound = async (action, ...params) => {
    try {
      return await action(...params)
    } catch (e) {
      if (!isNotFound(e)) {
        throw e
      }
    }

    return undefined
  }

  const results = await Promise.all([
    ignoreNotFound(tts.Applications.Devices.Repository.getUplinkDecoder, ...),
    ignoreNotFound(tts.Applications.Devices.Repository.getDownlinkDecoder, ...),
    ignoreNotFound(tts.Applications.Devices.Repository.getDownlinkEncoder, ...),
  ])
  
  return {
    ...results[0],
    ...results[1],
    ...results[2],
  }
```

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 14]
```diff
@@ -197,13 +204,46 @@ class PayloadFormattersForm extends React.Component {
...
+      return () =>
+        this.formRef?.current?.setFieldValue(FIELD_NAMES.JAVASCRIPT, applicationFormatter)
+    }
+
+    return () =>
+      this.formRef?.current?.setFieldValue(
+        FIELD_NAMES.JAVASCRIPT,
+        repoFormatters.formatter_parameter,
+      )
+  }
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]: Mind the complexity/readability/maintainability costs of making unified handlers vs explicitly writing distinct handlers:

```javascript
  @bind
  pasteAppPayloadFormatter() {
    const { defaultParameter, uplink } = this.props
    this.formRef?.current?.setFieldValue(FIELD_NAMES.JAVASCRIPT, defaultParameter || getDefaultJavascriptFormatter(uplink))
  }

  @bind
  pasteRepoPayloadFormatter() {
    const { repoFormatters } = this.props
    this.formRef?.current?.setFieldValue(
      FIELD_NAMES.JAVASCRIPT,
      repoFormatters.formatter_parameter,
    )
  }
```

Don't sacrifice code clarity for a tiny memory improvement (if any).

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 15]
```diff
@@ -197,13 +204,46 @@ class PayloadFormattersForm extends React.Component {
     }
   }
 
+  @bind
+  pastePayloadFormatter(app) {
+    const { defaultParameter, uplink, repoFormatters } = this.props
+    const applicationFormatter = defaultParameter
+      ? defaultParameter
+      : getDefaultJavascriptFormatter(uplink)
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
    const applicationFormatter = defaultParameter || getDefaultJavascriptFormatter(uplink)
```

You can just use a logical OR to achieve the same thing.

File: pkg/webui/console/components/payload-formatters-form/index.js (hunk 7 above)
Comment (kschiffer) [reply pkg/webui/console/components/payload-formatters-form/index.js]: Again, there is no reason to not hardcode the `readOnly` value as `true` here. The displayed repo formatter will always be read-only.

File: pkg/webui/console/components/payload-formatters-form/index.js [hunk 16]
```diff
@@ -261,6 +336,7 @@ class PayloadFormattersForm extends React.Component {
         initialType === TYPES.GRPC ? initialParameter : getDefaultGrpcServiceFormatter(uplink),
     }
     const options = allowReset ? formatterOptionsWithReset : formatterOptions
+    const defaultFormatter = defaultType.replace('FORMATTER_', '').toLowerCase()
```
Comment (kschiffer) [inline pkg/webui/console/components/payload-formatters-form/index.js]:
```suggestion
    const defaultFormatter = defaultType.replace('FORMATTER_', '').toLowerCase()
```

Not a fan of solutions like these. It abuses our internal naming, which is completely arbitrary, as user-facing communication. Someone working on this at a later stage might not be aware of this liability and choose namings that are not OK for being displayed to users. Additionally, these strings are not internationalized and only work in English.

A better solution would be to create a mapping of formatter types to i18n messages and use that instead, similar to what we are doing with `console/lib/field-description-messages.js`.

File: pkg/webui/console/store/middleware/logics/devices.js [hunk 17]
```diff
@@ -34,7 +34,7 @@ const getDeviceLogic = createRequestLogic({
     } = action
     const dev = await tts.Applications.Devices.getById(appId, deviceId, selector)
     dispatch(devices.startDeviceEventsStream(dev.ids))
-    return dev
+    return { ...dev }
```
Comment (kschiffer) [inline pkg/webui/console/store/middleware/logics/devices.js]:
```suggestion
    return dev
```

File: pkg/webui/console/store/reducers/link.js [hunk 18]
```diff
@@ -37,9 +37,8 @@ const getLinkFailure = (state, { payload }) => ({
   link: payload.link || {},
 })
 
-const updateLinkSuccess = (state, { link }) => {
-  const newLink = { ...state.link, ...link }
-
+const updateLinkSuccess = (state, link) => {
+  const newLink = { ...state.link, ...link.payload }
```
Comment (kschiffer) [inline pkg/webui/console/store/reducers/link.js:41]: What's the rationale behind this change? Has this been broken all along?

**Review by kschiffer — CHANGES_REQUESTED** [review]

File: pkg/webui/console/store/reducers/link.js (hunk 18 above)
Comment (ryaplots) [reply pkg/webui/console/store/reducers/link.js:41]: It was broken all along indeed. `{ link }` came as undefined every time. At least when I was testing. This is the reason behind that [support issue](https://github.com/TheThingsIndustries/lorawan-stack-support/issues/691)

File: pkg/webui/console/components/payload-formatters-form/index.js (hunk 7 above)
Comment (ryaplots) [reply pkg/webui/console/components/payload-formatters-form/index.js]: This repoFormatter will also we shown when JavaScript type is selected. Shouldn't the user be able to edit it then? I thought the idea was that it is only read-only when the default type is JS and when the type is Repository. So there is one case when it should not be read-only - when the chosen type is JavaScript and then the user can paste and appFormatter or repoFormatter and adjust it as they wish. For that the editor cannot be read-only. Or did I misunderstand something?

File: pkg/webui/console/containers/devices-table/index.js [hunk 19]
```diff
@@ -172,7 +172,13 @@ class DevicesTable extends React.Component {
       getDevicesList(
         props.appId,
         filters,
-        ['name', 'application_server_address', 'network_server_address', 'join_server_address'],
+        [
+          'name',
+          'application_server_address',
+          'network_server_address',
+          'join_server_address',
+          'version_ids',
```
Comment (kschiffer) [inline pkg/webui/console/containers/devices-table/index.js]: The list request is not the right place to source the version ids, since it is only executed when accessing a device via the device table. When accessing an end device directly (e.g. by clicking on a link that goes to a specific end device), the version information will be missing since the list request did not run. To get the version IDs safely, they need to be sourced from the single device request in `pkg/webui/console/views/device/index.js`.

File: pkg/webui/console/store/middleware/logics/device-repository.js [hunk 20]
```diff
@@ -84,10 +85,44 @@ const getTemplateLogic = createRequestLogic({
...
+      try {
+        return await action(...params)
+      } catch (e) {
+        if (!isNotFoundError(e)) {
+          throw e
+        }
+      }
+
+      return undefined
+    }
```
Comment (kschiffer) [inline pkg/webui/console/store/middleware/logics/device-repository.js]: It's not necessary to reassign this function every time the logic runs. You can rather define the function outside of the logic scope.

**Comment by ryaplots** [comment]
Button spacing now looks like:
Screenshot 2022-02-22 at 11 41 15

**Review by kschiffer — APPROVED** [review]
